perun-core/src/test/java/cz/metacentrum/perun/core/entry/VosManagerEntryIntegrationTest.java

```diff
@@ -775,6 +778,75 @@ public void addExistingMemberVo() throws Exception {
 			.isThrownBy(() -> vosManagerEntry.addMemberVo(sess, vo, vo2));
 	}
 
+	@Test
+	public void addMemberVoNoExistingMember() throws Exception {
+		System.out.println(CLASS_NAME + "addMemberVoNoExistingMember");
+
+		Vo vo = perun.getVosManagerBl().createVo(sess, myVo);
+		Vo memberVo = perun.getVosManagerBl().createVo(sess, new Vo(-1, "Vo2", "vo2"));
+
+		Member member = createMemberFromExtSource(memberVo);
+		vosManagerEntry.addMemberVo(sess, vo, memberVo);
+
+		assertThat(perun.getMembersManagerBl().getMembers(sess, vo).size()).isEqualTo(1);
+		assertThat(perun.getMembersManagerBl().getMembers(sess, vo).get(0).getUserId()).isEqualTo(member.getUserId());
+
+		List<Member> membersInParentVo = perun.getMembersManagerBl().getMembers(sess, vo);
+		assertThat(membersInParentVo.size()).isEqualTo(1);
+		assertThat((ArrayList<String>) perun.getAttributesManagerBl().getAttribute(sess, membersInParentVo.get(0), A_MEMBER_DEF_MEMBER_ORGANIZATIONS).getValue())
+			.contains(memberVo.getShortName());
+	}
+
+	@Test
+	public void addMemberVoExistingMember() throws Exception {
+		System.out.println(CLASS_NAME + "addMemberVoExistingMember");
+
+		Vo vo = perun.getVosManagerBl().createVo(sess, myVo);
+		Vo memberVo1 = perun.getVosManagerBl().createVo(sess, new Vo(-1, "Vo2", "vo2"));
+		Vo memberVo2 = perun.getVosManagerBl().createVo(sess, new Vo(-2, "Vo3", "vo3"));
+
+		Member member1 = createMemberFromExtSource(memberVo1);
+		Member member2 = perun.getMembersManagerBl().createMember(sess, memberVo2, perun.getUsersManagerBl().getUserByMember(sess, member1));
+		vosManagerEntry.addMemberVo(sess, vo, memberVo1);
+		vosManagerEntry.addMemberVo(sess, vo, memberVo2);
+
+		List<Member> membersInParentVo = perun.getMembersManagerBl().getMembers(sess, vo);
+		assertThat(membersInParentVo.size()).isEqualTo(1);
+		assertThat((ArrayList<String>) perun.getAttributesManagerBl().getAttribute(sess, membersInParentVo.get(0), A_MEMBER_DEF_MEMBER_ORGANIZATIONS).getValue())
+			.contains(memberVo1.getShortName(), memberVo2.getShortName());
+	}
+
+	@Test
+	public void removeMemberVoClearsMemberOrganizationsAttribute() throws Exception {
+		System.out.println(CLASS_NAME + "removeMemberVoClearsMemberOrganizationsAttribute");
+
+		Vo vo = perun.getVosManagerBl().createVo(sess, myVo);
+		Vo memberVo = perun.getVosManagerBl().createVo(sess, new Vo(-1, "Vo2", "vo2"));
+
+		Member member = createMemberFromExtSource(memberVo);
+		vosManagerEntry.addMemberVo(sess, vo, memberVo);
+
+		assertThat(perun.getMembersManagerBl().getMembers(sess, vo).size()).isEqualTo(1);
+
+		vosManagerEntry.removeMemberVo(sess, vo, memberVo);
+		assertThat((ArrayList<String>) perun.getAttributesManagerBl().getAttribute(sess, member, A_MEMBER_DEF_MEMBER_ORGANIZATIONS).getValue())
+			.isNullOrEmpty();
+	}
+
+	@Test
+	public void deleteVoInHierarchy() throws Exception {
```

Comment:
We should also probably test deleting the parent VO and it would be nice to have `RelationExistsException` use cases in tests as well.

Reply:
DeleteVo method re-throws all exceptions of existing relationships as InternalErrorException, so I added check that one is thrown if called with forceDelete == false.